src/Storages/FileLog/FileLogSource.cpp

```diff
@@ -0,0 +1,122 @@
+#include <DataStreams/ConvertingBlockInputStream.h>
+#include <DataStreams/OneBlockInputStream.h>
+#include <Formats/FormatFactory.h>
+#include <Interpreters/Context.h>
+#include <Processors/Executors/StreamingFormatExecutor.h>
+#include <Processors/Formats/InputStreamFromInputFormat.h>
+#include <Storages/FileLog/FileLogSource.h>
+#include <Storages/FileLog/ReadBufferFromFileLog.h>
+#include <Common/Stopwatch.h>
+#include <base/logger_useful.h>
+
+namespace DB
+{
+static constexpr auto MAX_FAILED_POLL_ATTEMPTS = 10;
+
+FileLogSource::FileLogSource(
+    StorageFileLog & storage_,
+    const StorageMetadataPtr & metadata_snapshot_,
+    const ContextPtr & context_,
+    const Names & columns,
+    size_t max_block_size_,
+    size_t poll_time_out_,
+    size_t stream_number_,
+    size_t max_streams_number_)
+    : SourceWithProgress(metadata_snapshot_->getSampleBlockForColumns(columns, storage_.getVirtuals(), storage_.getStorageID()))
+    , storage(storage_)
+    , metadata_snapshot(metadata_snapshot_)
+    , context(context_)
+    , column_names(columns)
+    , max_block_size(max_block_size_)
+    , poll_time_out(poll_time_out_)
+    , stream_number(stream_number_)
+    , max_streams_number(max_streams_number_)
+    , non_virtual_header(metadata_snapshot_->getSampleBlockNonMaterialized())
+    , virtual_header(
+          metadata_snapshot->getSampleBlockForColumns(storage.getVirtualColumnNames(), storage.getVirtuals(), storage.getStorageID()))
+{
+    buffer = std::make_unique<ReadBufferFromFileLog>(storage, max_block_size, poll_time_out, context, stream_number_, max_streams_number_);
+}
+
+FileLogSource::~FileLogSource()
+{
+    auto & file_infos = storage.getFileInfos();
+
+    size_t files_per_stream = file_infos.file_names.size() / max_streams_number;
+    size_t start = stream_number * files_per_stream;
+    size_t end = stream_number == max_streams_number - 1 ? file_infos.file_names.size() : (stream_number + 1) * files_per_stream;
+
+    /// Each stream responsible for close it's files and store meta
+    for (size_t i = start; i < end; ++i)
+    {
+        storage.closeFileAndStoreMeta(file_infos.file_names[i]);
+    }
```

Comment:
I think we should do this in `FileLogSource::onFinish()` instead. This will get read of that race.

Reply:
btw, I don't understand why this can get rid of that race, we still can start a new read(`select`) before last read pipeline finish?

Reply:
I was think we can pass a lock to FileLogSource?

Reply:
It wont, I did not read full stack trace.
Anyway moving this part to `onFinish` is more correct.
About the race.
Lets have a separate method in `FileLogStorage` -- `storeMeta` to which streams will go under mutex in an attempt to commit offsets. Also there should be a check that we store at offset which is bigger than the last one. And we can also copy files_info instead of reference.
Then the race I think will be fixed.
Also this case is not actually important because this storage is supposed to be used for MV only, SELECT queries are just to debug.

Reply:
There is a drawback in this idea, we actually do not need to lock the same mutex for different files.
May be you have something else in mind?
Also I think another solution is just to forbid select if there is mv. It is ok, since select is pointless for this srorage and is only for debug.

Reply:
> And we can also copy files_info instead of reference.

So we copy the full file_infos into each streams?

Reply:
Yes, not much expense I believe.

Reply:
Because we copy only file_infos which this stream needs.

Reply:
I realized that copy file_infos into streams does not work, because during the executing, the file_infos need to be updated.

Reply:
As I remember we update only last read offset, so this can be done in `storeMetadata` method.

Reply:
I made some change to update metadata after each written block. But since we still need to access and update file_infos onFinish, so I think we does not need to copy file_infos into steams.

Reply:
> But since we still need to access and update file_infos onFinish, so I think we don't need to copy file_infos into steams.

But it will generate races? Even though different files are processed by different sources, I am not sure it is ok.